Sources/AsyncAlgorithms/AsyncZipLatestFromSequence.swift

```diff
@@ -0,0 +1,382 @@
+  ///
+  /// ```
+  /// let base = AsyncChannel<Int>()
+  /// let other1 = AsyncChannel<String>()
+  /// let other2 = AsyncChannel<String>()
+  /// let sequence = base.zipLatest(from: other1, other2)
+  ///
+  /// Task {
+  ///   for element in await sequence {
+  ///    print(element)
+  ///   }
+  /// }
+  ///
+  /// await other1.send("a")
+  /// await other1.send("b")
+  ///
+  /// await other2.send("c")
+  /// await other2.send("d")
+  ///
+  /// ... later in the application flow
+  ///
+  /// await base.send(1)
+  ///
+  /// // will print: (1, "b", "d")
+  /// ```
+  ///
+  /// - Parameters:
+  ///   - other1: the first other ``AsyncSequence``
+  ///   - other2: the second other ``AsyncSequence``
+  /// - Returns: an ``AsyncZipLatestFrom2Sequence`` where elements are a tuple of an element from `self` and the
+  /// latest known elements (if any) from the `other` sequences.
+  @inlinable
+  func zipLatest<Other1: AsyncSequence, Other2: AsyncSequence>(
+    from other1: Other1,
+    _ other2: Other2
+  ) -> AsyncZipLatestFrom2Sequence<Self, Other1, Other2> {
+    AsyncZipLatestFrom2Sequence(self, other1, other2)
+  }
+}
+
+/// ``AsyncZipLatestFromSequence`` is an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+/// latest known element (if any) from the `other` sequence.
+public struct AsyncZipLatestFromSequence<Base: AsyncSequence, Other: AsyncSequence>: AsyncSequence where Other: Sendable {
+  public typealias Element = (Base.Element, Other.Element)
+  public typealias AsyncIterator = Iterator
+
+  @usableFromInline
+  let base: Base
+
+  @usableFromInline
+  let other: AsyncCurrentElementSequence<Other>
+
+  /// Creates an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+  /// latest known element (if any) from the `other` sequence.
+  /// - Parameters:
+  ///   - base: the ``AsyncSequence`` that drives the production of the output tuples
+  ///   - other: the ``AsyncSequence`` from which the latest known element is gathered to form the output tuple
+  @inlinable
+  public init(_ base: Base, _ other: Other) {
+    self.base = base
+    self.other = AsyncCurrentElementSequence(other)
+  }
+
+  @inlinable
+  public func makeAsyncIterator() -> Iterator {
+    Iterator(
+      self.base.makeAsyncIterator(),
+      self.other.makeAsyncIterator()
+    )
+  }
+
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var base: Base.AsyncIterator
+
+    @usableFromInline
+    var other: AsyncCurrentElementSequence<Other>.AsyncIterator
+
+    @usableFromInline
+    var isFinished = false
+
+    @usableFromInline
+    init(_ base: Base.AsyncIterator, _ other: AsyncCurrentElementSequence<Other>.AsyncIterator) {
+      self.base = base
+      self.other = other
+    }
+
+    @inlinable
+    public mutating func next() async rethrows -> Element? {
+      guard !Task.isCancelled else {
```

Comment:
so the cancellation only happens at the start of the iteration? it does not respond immediately while the base here is going too iterated (or when the other is iterated?

Reply:
If I am reading it correctly; I _think_ that is fine... just wanted to make sure that was the intent.

Reply:
I've reworked the cancellation handling so that every iteration can catch it